src/esp/gfx/RenderTarget.cpp

```diff
@@ -122,28 +161,41 @@ struct RenderTarget::Impl {
   void renderExit() {}
 
   void blitRgbaToDefault() {
-    if (rendererFlags_ & Renderer::Flag::NoTextures)
-      throw std::runtime_error(
```

Comment:
The throws instead of asserts here was intentional.  I expect that this is something python uses could reasonably end up hitting and a throw is much nicer to work with than an assert.  The assert exits hard while the throw can be caught in pdb and more easily debugged.

Reply:
@erikwijmans : 
- Within the current C++ simulator, we do not built exception handling mechanism (no "catch" there) so such throws will not be handled properly in C++. That means it actually fails "silently" (asssuming no gdb, pdb or any debugger, just by system itself.)
- Assertion actually saves debugging time. When a critical error happens, it is meaningless to let the simulator continues running. So if it is to fail, let it fail fast. "It caught quite a lot of mistakes that would take significant time to discover and debug otherwise". quoted from @mosra . I had a lengthy conversation on this topic with him when I wrote the PBR shader.  Somehow I agree with him now on this point. Read the discussion [here](https://github.com/facebookresearch/habitat-sim/pull/840#discussion_r504624575).
- "pdb can catch the throw". Yes, it might be. But the question is why the user would use pdb, gdb or any debugger first? The answer is she will not use it until something is wrong with the system, but the she cannot tell exactly where goes wrong. It starts costing her time to debug already. On the other hand, if the "system exits hard" with the proper error information, she already knows the location, and properly the root cause.

The essence is that to handle the error in a gentle way or a hard way. In the current system, most of the places choose the later so I decide to do the same to make it consistent.

Reply:
> Within the current C++ simulator, we do not build exception handling mechanism (no "catch" there) so such throws will not be handled properly in C++. That means it actually fails "silently" (asssuming no gdb, pdb or any debugger, just by system itself.)

There seems to be some miss-understanding here.  An uncaught exception will cause the program to exit immediately , not fail silently. What I mean by hard exit is that asserts cause the program to terminate in a way that pdb is not equipped to handle while in can handle cpp exceptions. 

I am by no means advocating that something fails silently.

Reply:
> terminate in a way that pdb is not equipped to handle while it can handle cpp exceptions

Ah, yes, without "try" and "catch", the program will exit. You are right. 

Now it behaves like the CORRADE_ASSERT (it will exit, and print out the error info). 
But CORRADE_ASSERT can give you the file name, line number where it fails, and the reasons. Now you already know where and why. May I understand why pdb is still needed?

Reply:
A reason, line number, and file name is nice but only really helpful if you already have an idea of what could go wrong -- awesome for devs, pretty good for power users, but not very useful for normal users.  With a throw you automatically call stack trace on the python side and debuggers are obviously nice to be able to use.

Reply:
OK. backtrace. I see your point.
Let us look at the case here. When an exception happens , it occurs at the rendering stage. So the backtrace will only give you a calling stack related to a serious of rendering calls. The genuine root cause, which is at the initialization stage, will not be shown in this backtrace at all. So I really do not see the benefits here. 

But you reached a very good point here, and we need to consider in the future development.

Reply:
What I do for Magnum python bindings is replicating asserts *outside* of the function and turning them into Python exceptions. So if a C++ developer "holds the function wrong", they get an assert, but if a Python dev does it, the bindings do all needed checks first and raise an exception *before* the actual C++ function gets called.

But I admit this is basically double the work :) (OTOH there's [CORRADE_NO_ASSERT](https://doc.magnum.graphics/corrade/Assert_8h.html#a588d67a03f847239f809eda035d73730) so, in case of magnum at least, the assert needs to be duplicated for python anyway.)

Reply:
My two cents: whatever we decide here (assert or exception), we should wrap it in a macro like HABITAT_RUNTIME_ERROR so we can make this decision in one place and document it.

The HABITAT_RUNTIME_ERROR macro is also a place to customize the behavior for other builds, e.g. the Javascript build.

Reply:
@erikwijmans : We heard you and we are actively seeking solutions to address your concerns and needs (Please see the docs in the slack channel for more details). Thanks.